src/environmentd/tests/sql.rs

```diff
@@ -671,75 +673,68 @@ fn test_subscribe_progress() {
     client_writes
         .batch_execute("CREATE TABLE t1 (data text)")
         .unwrap();
+    client_writes
+        .batch_execute("INSERT INTO t1 VALUES ('snapdata')")
+        .unwrap();
     client_reads
         .batch_execute(
             "COMMIT; BEGIN;
-         DECLARE c1 CURSOR FOR SUBSCRIBE t1 WITH (PROGRESS);",
+             DECLARE c1 CURSOR FOR SUBSCRIBE t1 WITH (PROGRESS)",
         )
         .unwrap();
 
-    #[derive(PartialEq)]
-    enum State {
-        WaitingForData,
-        WaitingForProgress(MzTimestamp),
-        Done,
+    // Asserts that the next data message is `data`. Ignores any progress
+    // messages that occur first.
+    //
+    // Returns the timestamp at which that data arrived.
+    fn await_data(client: &mut postgres::Client, data: &str) -> u64 {
+        // We have to try several times. It might be that the FETCH gets a a
+        // progress statements rather than data. We retry until we get the batch
+        // that has the data.
+        loop {
+            let rows = client
+                .query("FETCH 1 c1", &[])
+                .unwrap();
+            info!(row = ?rows.first());
+            let data_row = match rows.first() {
+                Some(row) if row.try_get::<_, String>("data").is_ok() => row,
+                _ => continue, // retry
+            };
+            assert_eq!(data_row.get::<_, bool>("mz_progressed"), false);
+            assert_eq!(data_row.get::<_, i64>("mz_diff"), 1);
+            assert_eq!(data_row.get::<_, String>("data"), data);
+            return data_row.get::<_, MzTimestamp>("mz_timestamp").0;
+        };
     }
 
+    // Asserts that the next progress message is at least `ts`.
```

Comment:
```suggestion
    // Asserts that the next message is a progress of at least `ts`.
```

Hard to come up with good wording, but the comment should document that it fails on data messages.